### PR TITLE
feature(zjow): Add Log Module the 'Traffic'.

### PR DESCRIPTION
**Pull request by zjowowen**

### CODE REVIEW

File: ding/utils/traffic_helper.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        Arguments:
+            - file_path (:obj:`str`): File path that offline data are to be saved at. \
+                In local mode, it must be set sucessfully for at least once. \
+                In remote mode, no need to set it for worker, but is necessary for master. \
+            - online (:obj:`bool`): To enable online analysis and maintain data in memory.
+            - maxlen (:obj:`int`): Max data size in memory.
+            - router (:obj:`Parallel`): To enable remote mode.
+        """
+
+        if online and maxlen <= 0:
```
Comment (sailxjx) [inline ding/utils/traffic_helper.py:39]: Use assert

File: ding/utils/traffic_helper.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            logging.error("Illegal data size.")
+            return self
+
+        if not hasattr(self, "_file") or self._file is None:
+            self._file = None
+            self._has_file_writer = False
+            self._file_path = None
```
Comment (sailxjx) [inline ding/utils/traffic_helper.py:46]: Why do you need so many properties for one feature?

File: ding/utils/traffic_helper.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                    if not os.path.exists(os.path.dirname(file_path)):
+                        os.makedirs(os.path.dirname(file_path))
+                    self._file = open(file_path, "a+")
+                    self._has_file_writer = True
+                    self._file_path = file_path
+                except IOError as error:
+                    logging.error("Invalid file path.")
+                    raise
+
+        if not hasattr(self, "_router") or self._router is None:
```
Comment (sailxjx) [inline ding/utils/traffic_helper.py:58]: Set these properties to None in __init__

File: ding/utils/traffic_helper.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            self._in_parallel = False
+            self._router = None
+            if router and router.is_active:
+                self._in_parallel = True
+                self._router = router
+                if self._has_file_writer:
+                    router.on("_Traffic_", self._on_msg)
+
+        if not hasattr(self, "_has_data") or self._has_data is False:
+            self._has_data = False
```
Comment (sailxjx) [inline ding/utils/traffic_helper.py:68]: What is this used for?

File: ding/utils/traffic_helper.py (hunk 1 above)
Comment (zjowowen) [reply ding/utils/traffic_helper.py]: fixed.

File: ding/utils/traffic_helper.py (hunk 2 above)
Comment (zjowowen) [reply ding/utils/traffic_helper.py]: Variables are reduced.

File: ding/utils/traffic_helper.py (hunk 3 above)
Comment (zjowowen) [reply ding/utils/traffic_helper.py]: Fixed.

File: ding/utils/traffic_helper.py (hunk 4 above)
Comment (zjowowen) [reply ding/utils/traffic_helper.py]: Fixed.

Comment (zjowowen) [reply ding/utils/traffic_helper.py]: It is an outdated feature for analysis. Now that it is no longer needed.

File: ding/utils/traffic_helper.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        Arguments:
+            - file_path (:obj:`str`): File path that offline data are to be saved at. \
+                In local mode, it must be set sucessfully for at least once. \
+                In remote mode, no need to set it for worker, but is necessary for master. \
+            - online (:obj:`bool`): To enable online analysis and maintain data in memory.
+            - maxlen (:obj:`int`): Max data size in memory.
+            - router (:obj:`Parallel`): To enable remote mode.
+        """
+
+        if online:
```
Comment (hiha3456) [inline ding/utils/traffic_helper.py:41]: what's the meaning of this part of code?

Comment (zjowowen) [reply ding/utils/traffic_helper.py:41]: To ensure parameter that given by user is reasonable.

File: ding/utils/traffic_helper.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            if kwargs:
+                dict_to_record = {**kwargs}
+            else:
+                dict_to_record = {}
+
+        if "__time" not in dict_to_record:
+            dict_to_record.update({"__time": time.time()})
+
+        msg = json.dumps(dict_to_record)
+        if self._file:
```
Comment (hiha3456) [inline ding/utils/traffic_helper.py:96]: maybe using elif here?

Comment (zjowowen) [reply ding/utils/traffic_helper.py:96]: A good suggestion!

**Pull request closed without merging**